docs/reference/cmodules.rst

```diff
@@ -0,0 +1,89 @@
+
+`Unlike CPython <https://docs.python.org/3/extending/building.html>`_, these
+modules are (currently) embedded directly in the program image instead of being
+dynamically loaded. This requires a `custom build of MicroPython
+<https://github.com/micropython/micropython/wiki/Getting-Started>`_.
+
+
+Writing a module
+----------------
+
+A module is a directory with the following files:
+
+  * ``micropython.mk``, which contains the Makefile fragment for this module.
+  * All C files you would like included.
+
+Put the required build commands in ``micropython.mk``. For a simple module, you
+will only have to add the file paths to ``SRC_MOD``, which will include these C
+files in the build:
+
+.. highlight:: make
+.. code::
+
+    # Add all C files to SRC_MOD.
+    SRC_MOD += $(MODULE_DIR)/example/example.c
+
+This is a very bare bones module named ``example`` that provides
+``example.double(x)``. Note that the name of the module must be equal to the
+directory name and is also used in the name of the ``mp_obj_module_t`` object at
+the bottom.
+
+.. highlight:: c
+.. code::
+
+    // Include required definitions first.
+    #include "py/obj.h"
+    #include "py/runtime.h"
+
+    // This is the function you will call using example.double(n).
+    STATIC mp_obj_t example_double(mp_obj_t x_obj) {
+        // Check input value and convert it to a C type.
+        if (!MP_OBJ_IS_SMALL_INT(x_obj)) {
+            mp_raise_ValueError("x is not a small int");
+        }
+        int x = mp_obj_int_get_truncated(x_obj);
+
+        // Calculate the double, and convert back to MicroPython object.
+        return mp_obj_new_int(x + x);
+    }
+    STATIC MP_DEFINE_CONST_FUN_OBJ_1(example_double_obj, example_double);
+
+    // Define all properties of the example module, which currently are the name (a
+    // string) and a function.
+    // All identifiers and strings are written as MP_QSTR_xxx and will be
+    // optimized to word-sized integers by the build system (interned strings).
+    STATIC const mp_rom_map_elem_t example_module_globals_table[] = {
+        { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_example) },
+        { MP_ROM_QSTR(MP_QSTR_double), MP_ROM_PTR(&example_double_obj) },
+    };
+    STATIC MP_DEFINE_CONST_DICT(example_module_globals, example_module_globals_table);
+
+    // Define module object.
+    const mp_obj_module_t example_user_cmodule = {
+        .base = { &mp_type_module },
+        .globals = (mp_obj_dict_t*)&example_module_globals,
+    };
+
+
+Using a module
+--------------
+
+To build such a module, compile MicroPython (see `getting started
+<https://github.com/micropython/micropython/wiki/Getting-Started>`_) with an
+extra ``make`` flag named ``USER_C_MODULES`` to the directory containing all
+modules you want included (not to the module itself!). For example, for the
+ESP8266 target:
+
+.. highlight:: shell
+.. code::
+
+    $ make axtls
+    $ make USER_C_MODULES=path-to-modules-folder
+    $ make deploy
```

Comment:
You don't need an explicit `make axtls` because that should be done automatically.  And with `make deploy` it's wise to keep the extra make options there, like `make USER_C_MODULES=path-to-modules-folder deploy`.  This is in case the make runs (part of) the build again before deploying, you don't want it to use different CFLAGS for certain parts of the build.

Instead of typing this long make command over and over again (and sometimes forgetting), it's easier to create a local `makefile` or `GNUmakefile` with your custom settings and then include the MicroPython `Makefile`.

Reply:
Note that if you don't have an explicit `make axtls` then these instructions become generic for any port.

Reply:
> You don't need an explicit make axtls because that should be done automatically

It's only done automatically in few (one?) ports. (E.g. not done in unix port).

Reply:
I've removed the ESP8266-specific parts. The only goal of it was showing how the make flag works (for those not that familiar with make), not as a guide for how to deploy.

It may be useful to write something about make flags and `GNUmakefile` files in a compilation guide, but I think that's outside of the scope of this PR.

> It's only done automatically in few (one?) ports. (E.g. not done in unix port).

Not in the ESP8266 port either, while I seem to remember it worked before. It would be nice if this were fixed, though. I have so many times run `make` just to discover I had to run `make axtls` first. Again, that's outside the scope of this PR.